go/vt/vtgate/planbuilder/operators/route_planning.go

```diff
@@ -80,15 +80,25 @@ func buildVindexTableForDML(
 	ctx *plancontext.PlanningContext,
 	tableInfo semantics.TableInfo,
 	table *QueryTable,
+	stmt sqlparser.Statement,
 	dmlType string,
 ) (*vindexes.Table, Routing) {
 	vindexTable := tableInfo.GetVindexTable()
-	if vindexTable.Source != nil {
+	if tableInfo.GetVindexTable().Type == vindexes.TypeReference && vindexTable.Source != nil {
 		sourceTable, _, _, _, _, err := ctx.VSchema.FindTableOrVindex(vindexTable.Source.TableName)
 		if err != nil {
 			panic(err)
 		}
 		vindexTable = sourceTable
+		refTbl := sqlparser.NewAliasedTableExpr(vindexTable.GetTableName(), "")
+		switch stmt := stmt.(type) {
+		case *sqlparser.Update:
+			stmt.TableExprs[0] = refTbl
+		case *sqlparser.Insert:
+			stmt.Table = refTbl
+		default:
+			panic("unsupported DML type in buildVindexTableForDML")
```

Comment:
we should not panic here and return vterrors.VT13001

Reply:
As discussed, this function doesn't return an error and the standard pattern used is to panic here and catch panics at an outer level.